python/ray/tune/integration/mlflow.py

```diff
@@ -0,0 +1,337 @@
+class MLFlowLoggerCallback(LoggerCallback):
+    """MLFlow Logger to automatically log Tune results and config to MLFlow.
+
+    MLFlow (https://mlflow.org) Tracking is an open source library for
+    recording and querying experiments. This Ray Tune ``LoggerCallback``
+    sends information (config parameters, training results & metrics,
+    and artifacts) to MLFlow for automatic experiment tracking.
+
+    Args:
+        tracking_uri (str): The tracking URI for where to manage experiments
+            and runs. This can either be a local file path or a remote server.
+            This arg gets passed directly to mlflow.tracking.MlflowClient
+            initialization. When using Tune in a multi-node setting, make sure
+            to set this to a remote server and not a local file path.
+        registry_uri (str): The registry URI that gets passed directly to
+            mlflow.tracking.MlflowClient initialization.
+        experiment_name (str): The experiment name to use for this Tune run.
+            If the experiment alread exists with Mlflow, it will be reused.
+            If not, a new experiment will be created.
+        save_artifact (bool): If set to True, automatically save the entire
+            contents of the Tune local_dir as an artifact to the
+            corresponding run in Mlflow.
+
+    Example:
+
+    .. code-block:: python
+
+        from ray.tune.integration.mlflow import MLFlowLoggerCallback
+        tune.run(
+            train_fn,
+            config={
+                # define search space here
+                "parameter_1": tune.choice([1, 2, 3]),
+                "parameter_2": tune.choice([4, 5, 6]),
+            },
+            callbacks=[MLFlowLoggerCallback(
+                experiment_name="experiment1",
+                save_artifact=True)])
+
+    """
+
+    def __init__(self,
+                 tracking_uri: Optional[str] = None,
+                 registry_uri: Optional[str] = None,
+                 experiment_name: Optional[str] = None,
+                 save_artifact: bool = False):
+        if mlflow is None:
+            raise RuntimeError("mlflow has not been installed. Please `pip "
+                               "install mlflow` to use the MLFlowLogger.")
+
+        self.client = MlflowClient(
+            tracking_uri=tracking_uri, registry_uri=registry_uri)
+
+        # Setup MLFlow experiment.
+        experiment_id = None
+        if experiment_name is None:
+            # If no name is passed in, then check env vars.
+            # First check if experiment_name env var is set.
+            experiment_name = os.environ.get("MLFLOW_EXPERIMENT_NAME")
+            if experiment_name is None:
+                # Then check if experiment_id is set and exists.
+                experiment_id = int(os.environ.get("MLFLOW_EXPERIMENT_ID"))
+                if experiment_id is None or not self.client.get_experiment(
+                        experiment_id):
+                    experiment_id = None
```

Comment:
This is a bit confusing, can we do this without so many nested ifs? E.g. first set `experiment_name` from the ENV var in any case, set the ID from the ENV var in any case, and set to None only if `get_experiment` fails

Reply:
Yes good point. I refactored it to make it clearer.